### ISSUE TITLE
Terminal processes on Windows (conhost.exe) do not get destroyed/reused for build tasks

### ISSUE THREAD

**Issue by Venryx** [microsoft/vscode#86472]
Issue Type: <b>Bug</b>

### Reproduction instructions

1) Clone this repo: https://github.com/Venryx/vscode-terminal-bug

2) Open the repo's root folder in VSCode, on a Windows machine. (it's possible it's an issue on other OS'es as well, but I'm guessing not, since otherwise it'd probably be fixed by now -- it's plagued me for months)

3) Press Help -> Open Process Explorer, and leave the window open to the side (so you can see the phantom conhost processes).

4) Run the "Tasks: Run Build Task" command, and select one of the two "build tasks" (they're just endless echo loops). You'll see new processes in the process-explorer.

5) Run the "Tasks: Terminate Task" command, and "terminate" the just-started task. The process-count will **not** go down, I suppose because vscode is trying to reuse the terminal processes, as per the message "Terminal will be reused by tasks, press any key to close it." when terminating using ctrl+c manually.

6) However, if you run the "Tasks: Run Build Task" command again, the old conhost processes do not get reused. Instead, they just stay around as phantom processes.

7) If you terminate the second build task, and start it again, yet another instance of conhost starts, with none of the old conhosts getting removed/reused. And so on, for every new launch of a build task.

Expected results: After starting and terminating the "cmd-loop" build task 3 times (as described above), I expect there to at most be 1 "terminal process-tree" left remaining for that build task.

Actual results: After 3 starts and termination, 3 copies of the terminal process-tree are left remaining. (as seen in the process-explorer, as well as third-party process explorers like Process Hacker 2)

### Screen capture

Here is a screen capture of me performing the steps above on a fresh clone of this repo:
![](https://github.com/Venryx/vscode-terminal-bug/raw/master/ScreenCapture.gif)

Note: The capture above is with extensions enabled, but I redid the steps with all extensions disabled, and the exact same output occurs.

### Why this is a problem

While the "phantom terminal processes" are an issue in and of themselves, the reason I took the time to work through all of this is that I located it as the root cause for a different, more severe bug I've recently encountered.

However, for this second, more severe bug:
1) It appears to be specific to my hardware (I think it started happening after I changed by GPU from a Nvidia 980 TI to an AMD RX Vega 56).
2) It appears to be specific to my build task (running webpack -- not sure if specific to my exact config).

Because of those two "context-specific" conditions, I haven't taken the time to document that bug yet, since it would be hard for the vscode developers to fix it. (since they can't as easily reproduce it)

That said, the short version of that bug is that, whenever a phantom terminal process is left behind (as seen above), and I start another build task, a few seconds later the entire vscode program freezes -- permanently. The only way to fix it is to force-kill the old phantom conhost.exe process, at which point vscode instantly unfreezes.

So for this second bug, I assume it has some sort of lock condition where one "terminal slot" is only supposed to have one conhost process attached to it, but because the old one doesn't get destroyed, two end up coexisting, and they cause a freeze when they both "communicate with that slot". I don't know the internals, but that's my end-user guess at what is happening. (however, it's more complicated than this, since the freeze only happens when I run my actual project's "webpack" build task, not when I run the simple echo-loop as seen above)

### Conclusion

Anyway, I may create a separate issue report for that freeze issue later, but for now I just want to get my project working again, and I think a fix to this phantom-conhost-process issue (which is easy to reproduce) would be sufficient for that.

---

VS Code version: Code 1.40.2 (f359dd69833dd8800b54d458f6d37ab7c78df520, 2019-11-25T14:54:45.096Z)
OS version: Windows_NT x64 10.0.18363
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 7 1700X Eight-Core Processor          (16 x 3400)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off<br>surface_control: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.95GB (11.19GB free)|
|Process Argv|--disable-extensions|
|Screen Reader|no|
|VM|0%|
</details>Extensions disabled

**Comment by Tyriar** [comment]
Not yet confirmed

**Comment by NotWearingPants** [comment]
I can confirm this happens to me as well when using nodemon [like the docs say](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_restarting-debug-sessions-automatically-when-source-is-edited):

```json
{
  "name": "Launch server.js via nodemon",
  "type": "node",
  "request": "launch",
  "runtimeExecutable": "nodemon",
  "program": "${workspaceFolder}/server.js",
  "restart": true,
  "console": "integratedTerminal",
  "internalConsoleOptions": "neverOpen"
}
```

The docs even mention it:

> Tip: Pressing the Stop button stops the debug session and disconnects from Node.js, but nodemon (and Node.js) will continue to run. To stop nodemon, you will have to kill it from the command line (which is easily possible if you use the integratedTerminal as shown above).

So it seems like it is a known issue, and the docs themselves confirm it.

**Comment by Venryx** [comment]
This is still an issue in VSCode 1.51.0:
```
Version: 1.51.0 (system setup)
Commit: fcac248b077b55bae4ba5bab613fd6e9156c2f0c
Date: 2020-11-05T18:18:23.642Z
Electron: 9.3.3
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Windows_NT x64 10.0.19041
```

If you're on Windows, and you:
1) Kill a build task. (using `Tasks > Terminate Task` command)
2) Start a new build task.

Then there are still "phantom conhost processes" left behind from the first build-task execution. (see original post for more details)

If that were all that happened, it would not be so bad. But, on my computer at least, the existence of these "phantom conhost processes" causes the entire VSCode window to freeze up (on starting a new build task).

This happens consistently, and it continues happening even after I changed my GPU from the AMD RX Vega 56 to the RTX 2060 Super (I previously wrote I thought the freeze was GPU related, but apparently not).

This is quite annoying as it makes the `Tasks > Terminate Task` command unusable for me. (instead I have to kill the terminal process manually each time with ctrl+c; and this is not possible in all cases, requiring me to then track down the phantom processes in Task Manager / Process Hacker 2 to kill them manually)

**Comment by Tyriar** [comment]
@Venryx fyi that version is like 6 months old

**Comment by Venryx** [comment]
This _seems_ to finally be resolved, in the latest versions of vscode (v1.58.2+). (if true, hurray!)

Can anyone else who had this issue confirm/deny?

**Comment by Tyriar** [comment]
@NotWearingPants do you still see this?
